cli/js/permissions.ts

```diff
@@ -53,25 +55,84 @@ export type PermissionDescriptor =
   | PluginPermissionDescriptor
   | HrtimePermissionDescriptor;
 
-export class PermissionStatus {
-  constructor(public state: PermissionState) {}
-  // TODO(kt3k): implement onchange handler
+interface StatusCacheValue {
+  state: PermissionState;
+  status: PermissionStatus;
+}
+
+export class PermissionStatus extends EventTarget {
+  #state: { state: PermissionState };
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  onchange: ((this: PermissionStatus, ev: Event) => any) | null = null;
+
+  get state(): PermissionState {
+    return this.#state.state;
+  }
+
+  constructor(state: { state: PermissionState }) {
+    super();
+    this.#state = state;
+  }
+
+  dispatchEvent(evt: Event): boolean {
+    let dispatched = super.dispatchEvent(evt);
+    if (dispatched && this.onchange) {
+      this.onchange.call(this, evt);
+      dispatched = !evt.defaultPrevented;
+    }
+    return dispatched;
+  }
+
+  get [Symbol.toStringTag](): string {
+    return "PermissionStatus";
+  }
+}
+
+/** A cache of `PermissionStatus` objects and their last known state. */
+const statusCache = new Map<string, StatusCacheValue>();
+
+/** Cache the state of a descriptor and return its `PermissionStatus`. */
+function cache(
+  desc: PermissionDescriptor,
+  state: PermissionState
+): PermissionStatus {
+  let key = desc.name;
+  if ((desc.name === "read" || desc.name === "write") && desc.path) {
+    key += `-${desc.path}`;
+  } else if (desc.name === "net" && desc.url) {
+    key += `-${desc.url}`;
+  }
+  if (statusCache.has(key)) {
+    const status = statusCache.get(key)!;
+    if (status.state !== state) {
+      status.state = state;
+      status.status.dispatchEvent(new Event("change", { cancelable: false }));
```

Comment:
If there's no cached entry then this event won't fire on first change?

Reply:
Incorrect.  It get cached before it is possible for it to change.  Every time a descriptor is sent, cache is called, it either creates a new descriptor or it returns the cached one, and if the current state is different than the cached state, it fires the event.  Every change fires the event.